### PR TITLE
Allow Chatbot examples to show more than one image

### PR DESCRIPTION
**Pull request by hannahblair**
## Description

This PR allows the Chatbot to display more files and info in the examples area and adds some design tweaks. Let me know what you think!

Screenshot 2024-12-05 at 00 28 51 

Screenshot 2024-12-05 at 00 28 58

--- 
Small question, should we cascade the image icons as well like we do for the files? 🤔 

Closes: #10111

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Testing and Formatting Your Code

1. PRs will only be merged if tests pass on CI. We recommend at least running the backend tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the backed tests: `bash scripts/run_backend_tests.sh`

2. Please run these bash scripts to automatically format your code: `bash scripts/format_backend.sh`, and (if you made any changes to non-Python files) `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by abidlabs** [comment]
Wow this looks splendid @hannahblair!

My one hesitation is the addition of the "Aa" icon on top of examples without any files. I do like that it makes all of the examples the same vertical height. However, it feels quite unnecessary when all of the examples only consist of text, which is probably the most common case (e.g. see `demo/chatinterface_artifacts` or `demo/chatinterface_multimodal`).

Idea: what if we only show this "Aa" icon if `multimodal=True`? Examples with files are more likely to occur when `multimodal=True`, in which having this icon is nice to have for the sake of clarity. 

> Small question, should we cascade the image icons as well like we do for the files? 🤔

I don't think we should cascade the image icons. For many VLMs, examples will often consist of a pair of images along with a question like, "what's different between these images" or something, in which case, it'll help to be able to see the images clearly before clicking on the example.

For the sake of consistency, I'd suggest we also not cascade the file icons, but open to other folks thoughts about both of these points. cc @aliabid94 @dawoodkhan82 @allisonwhilden

**Comment by dawoodkhan82** [comment]
This is clean!

Agree on not cascading the file icons. Also maybe we can show a different icon for audio and video files, like we do in the multimodal textbox. Or even nicer would be to show a thumbnail for video files.

**Comment by abidlabs** [comment]
Agree with @dawoodkhan82 about icons for audio and video files, it'd be good to have consistent icons in the chatbot examples and the multimodal textbox:

image

**Comment by abidlabs** [comment]
Not sure if this PR is 100% ready @hannahblair but I noticed that there's a gap on top of the the examples when `mutlimodal=False`:

image

(see `demo/chatinterface_multimodal/run.py`)

**Comment by abidlabs** [comment]
@hannahblair not sure if I'm testing wrong, but I just tested with this modified version of `chatinterface_multimodal/run.py` and I can't see any of the file icons in the examples:

```py
import gradio as gr

def echo(message, history):
    return message["text"]

demo = gr.ChatInterface(
    fn=echo,
    type="messages",
    examples=[{"text": "hello", "files": ["files/avatar.png"]}, {"text": "hola", "files": ["files/avatar.png", "files/cantina.wav"]}, {"text": "merhaba", "files": ["files/avatar.png", "files/avatar.png", "files/avatar.png"]}],
    title="Echo Bot",
    multimodal=True,
)
demo.launch()
```

image


Does this work for you?

**Comment by abidlabs** [comment]
Sorry I'm confused @hannahblair why are adding the `multimodal` attribute to `gr.Chatbot`? Why does it need to know whether or not the ChatInterface is multimodal or not?

**Comment by abidlabs** [comment]
Ah okay so I understand the reason for creating the `multimodal ` parameter now, but I don't think we should use this approach because its a leaky abstraction i.e. there's no reason the `gr.Chatbot` component should need to know about any higher-level abstraction like `gr.ChatInterface` or the `gr.MultimodalTextbox`. 

There's a couple of approaches we could take instead:

(1) The `gr.ChatInterface` could be responsible for setting the icon of text-only examples before they are passed into `gr.Chatbot`. In other words, if the `gr.ChatInterface` is multimodal and there's a text-only example that does not have an example_icon, its example_icon automatically gets set to "Aa" icon. Then, its passed into `gr.Chatbot`, which renders the example as it would normally, including the "Aa" example_icon.

(2) The other possibility would to modify the logic a little bit so that instead of showing the "Aa" icon when the gr.ChatInterface is multimodal, we show it for any text-only example _if there are any other examples that include files_. This logic could be contained entirely inside `gr.Chatbot` since it knows all of the examples. 

My preference is for (1) since it feels less "surprising", but (2) also makes sense as long as we document the behavior appropriately.

**Comment by abidlabs** [comment]
> This won't work without some modification in the backend because even when multimodal=False in the demo, we still receive the examples with their files and icons if they've been passed to the Chat Interface. I'm thinking we need to strip the files from examples=self.examples_messages if multimodal=False when we pass the examples to Chatbot here:

Why do we need to strip the files? There might be some niche cases where a user really wants to pass in files in their example for a non-multimodal chat interface, can we just let them do that? I can't think of any reason for it, but at the same time, it doesn't break anything if they do. On the other hand, it might be a breaking change for certain demos that might be using this feature.

File: gradio/chat_interface.py [hunk 1]
```diff
@@ -338,8 +339,18 @@ def _setup_example_messages(
...
+                    if not example_files:
+                        example_message["icon"] = {
+                            "path": "",
+                            "url": None,
+                            "orig_name": None,
+                            "mime_type": "text",  # for internal use, not a valid mime type
+                            "meta": {"_type": "gradio.FileData"},
+                        }
+                    else:
+                        example_message["icon"] = example_icons[index]
```
Comment (abidlabs) [inline gradio/chat_interface.py:353]: if I understand correctly, should this be something like this:
```suggestion
                if multimodal:
                    example_files = example_message.get("files")
                    if not example_files:
                      if not not example_icons:
                        example_message["icon"] = {
                            "path": "",
                            "url": None,
                            "orig_name": None,
                            "mime_type": "text",  # for internal use, not a valid mime type
                            "meta": {"_type": "gradio.FileData"},
                        }
                      else:
                        example_message["icon"] = example_icons[index]
```

**Comment by abidlabs** [comment]
Gave this a quick spin before I log off and generally things are looking good, although I didn't see the "Aa" icon above the text-only example as I expected. Also not sure if multiple files are being rendered:

image



(This is running `demo/test_chatinterface_multimodal_examples/run.py`)

File: gradio/chat_interface.py [hunk 2]
```diff
@@ -338,8 +339,21 @@ def _setup_example_messages(
                     example_message["files"] = example.get("files", [])
                 if example_labels:
                     example_message["display_text"] = example_labels[index]
-                if example_icons:
-                    example_message["icon"] = example_icons[index]
+                if multimodal:
```
Comment (abidlabs) [inline gradio/chat_interface.py:342]:
```suggestion
                if self.multimodal:
```

Comment (abidlabs) [reply gradio/chat_interface.py:342]: I would instead do this. 

You'll also need to remove the `@staticmethod` decorator from the `_setup_example_messages` method.

**Review by abidlabs — APPROVED** [review]
lgtm @hannahblair looks great! Just left one minor suggestion on the backend implementation

**Pull request merged**